pkg/operator2/configobservation/routersecret/observe_router_secret.go

```diff
@@ -0,0 +1,91 @@
+	"github.com/openshift/library-go/pkg/operator/events"
+
+	"github.com/openshift/cluster-authentication-operator/pkg/operator2/configobservation"
+)
+
+func ObserveRouterSecret(genericlisters configobserver.Listers, recorder events.Recorder, previouslyObservedConfig map[string]interface{}) (ret map[string]interface{}, _ []error) {
+	namedCertificatesPath := []string{configobservation.OAuthServerConfigPrefix, "servingInfo", "namedCertificates"}
+	defer func() {
+		ret = configobserver.Pruned(ret, namedCertificatesPath)
+	}()
+
+	listers := genericlisters.(configobservation.Listers)
+	errs := []error{}
+
+	currentNamedCertificates, _, err := unstructured.NestedFieldCopy(previouslyObservedConfig, namedCertificatesPath...)
+	if err != nil {
+		// continue on read error from existing config in an attempt to fix it
+		errs = append(errs, err)
+	}
+
+	currentNamedCertificatesSlice := []interface{}{}
+	if currentNamedCertificates != nil {
+		currentNamedCertificatesSlice = currentNamedCertificates.([]interface{})
+	}
+
+	routerSecret, err := listers.SecretsLister.Secrets("openshift-authentication").Get("v4-0-config-system-router-certs")
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedNamedCertificates, err := routerSecretToSNI(routerSecret)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedConfig := map[string]interface{}{}
+	if err := unstructured.SetNestedField(
+		observedConfig,
+		observedNamedCertificates,
+		namedCertificatesPath...,
+	); err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	if !equality.Semantic.DeepEqual(currentNamedCertificatesSlice, observedNamedCertificates) {
+		recorder.Eventf("ObserveRouterSecret", "namedCertificates changed to %#v", observedNamedCertificates)
+	}
+
+	return observedConfig, errs
+}
+
+func routerSecretToSNI(routerSecret *corev1.Secret) ([]interface{}, error) {
+	var certs []configv1.NamedCertificate
+	// make sure the output slice of named certs is sorted by domain so that the generated config is deterministic
+	for _, domain := range sets.StringKeySet(routerSecret.Data).List() {
+		certs = append(certs, configv1.NamedCertificate{
+			Names: []string{"*." + domain}, // ingress domain is always a wildcard
+			CertInfo: configv1.CertInfo{ // the cert and key are appended together
+				CertFile: "/var/config/system/secrets/v4-0-config-system-router-certs/" + domain,
+				KeyFile:  "/var/config/system/secrets/v4-0-config-system-router-certs/" + domain,
+			},
+		})
+	}
+
+	configBytes, err := json.Marshal(certs)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: we may want to actually round-trip this to remove the fields that can be omitted
+	unstructuredCertConfig := []interface{}{}
+	if err := json.NewDecoder(bytes.NewBuffer(configBytes)).Decode(&unstructuredCertConfig); err != nil {
+		return nil, fmt.Errorf("decode of observed config failed with error: %v", err)
+	}
+
+	return unstructuredCertConfig, nil
```

Comment:
isn't it easier to just create the unstructured directly above than this encoding+decoding magic?